src/libxtp/threecenter_gwbse.cc

```diff
@@ -150,15 +176,55 @@ void TCMatrix_gwbse::FillBlock(std::vector<Eigen::MatrixXd>& block,
       }
     }  // gamma-loop
   }    // alpha-loop
-  for (int k = 0; k < auxshell.getNumFunc(); ++k) {
+
+#if defined(USE_GPU)
+  // create tensor
+  std::vector<Eigen::MatrixXd> tensor;
+  for (const Eigen::MatrixXd& mtx : symmstorage) {
+    tensor.push_back(mtx.selfadjointView<Eigen::Lower>());
+  }
+  // Performn the tensor multiplication in the GPU
+  try {
+    std::vector<Eigen::MatrixXd> results =
+        _gpu_handle.triple_tensor_product(dftn.transpose(), dftm, tensor);
+    for (int k = 0; k < auxshell.getNumFunc(); ++k) {
+      Eigen::MatrixXd threec_inMo = results[k];
+      for (int i = 0; i < threec_inMo.cols(); ++i) {
+        block[i].col(k) = threec_inMo.col(i);
+      }
+    }
+  } catch (const std::runtime_error& error) {
+    XTP_LOG_SAVE(logDEBUG, _log)
+        << TimeStamp()
+        << " GPU tripe_tensor_product failed due to: " << error.what()
+        << " Using default CPU TripleTensorProduct!" << flush;
+
+    this->TripleTensorProduct(block, symmstorage, dftn, dftm,
+                              auxshell.getNumFunc());
+  }
+#else
+  this->TripleTensorProduct(block, symmstorage, dftn, dftm,
+                            auxshell.getNumFunc());
+#endif
+  return;
+}
+
+/*
+ * Perform a Matrix Tensor Matrix multiplication
+ */
+void TCMatrix_gwbse::TripleTensorProduct(
+    std::vector<Eigen::MatrixXd>& block,
+    const std::vector<Eigen::MatrixXd>& symmstorage,
+    const Eigen::MatrixXd& dftn, const Eigen::MatrixXd& dftm,
+    int numFunc) const {
```

Comment:
you do not need numfunc, as it results from the other input parameters

Reply:
removed `numFunc`